client/src/main/java/se/fortnox/reactivewizard/client/RestClientFactory.java

```
@@ -4,41 +4,73 @@
+    private <T> Provider<T> provider(Class<T> iface,
+        Provider<HttpClientProvider> httpClientProvider, Provider<? extends HttpClientConfig> httpClientConfigProvider) {
         return () -> {
-            T httpProxy = httpClientProvider.get().create(iface);
-            log.debug("Created {} for {}", Proxy.getInvocationHandler(httpProxy), iface.getName());
+
+            HttpClientConfig httpClientConfig = httpClientConfigProvider.get();
+
+            //Create client based on config and create proxy
+            T httpProxy = httpClientProvider.get().createClient(httpClientConfig).create(iface);
+
+            LOG.debug("Created {} for {} with custom httpClient: {}", Proxy.getInvocationHandler(httpProxy), iface.getName(), httpClientConfig);
             return httpProxy;
         };
     }
 
     @SuppressWarnings({"unchecked", "rawtypes"})
     @Override
     public void configure(Binder binder) {
-        Multibinder.newSetBinder(binder, TypeLiteral.get(PreRequestHook.class));
-        Multibinder.newSetBinder(binder, TypeLiteral.get(RequestParameterSerializer.class));
-        Provider<HttpClient> httpClientProvider = binder.getProvider(HttpClient.class);
+        Multibinder.newSetBinder(binder, PreRequestHook.class);
+        Multibinder.newSetBinder(binder, RequestParameterSerializer.class);
+        Provider<HttpClientProvider> httpClientProvider = binder.getProvider(HttpClientProvider.class);
+
+        Map<Class, Class<?>> httpClientConfigByResource = new HashMap<>();
+
+        httpConfigClassScanner
+            .getClasses()
+            .forEach(configClass -> {
+                if (configClass.isAnnotationPresent(UseInResource.class)) {
+                    for (Class resource : configClass.getAnnotation(UseInResource.class).value()) {
+                        if (!resource.isInterface()) {
+                            LOG.warn("{} pointed out in UseInResource must be an interface", resource);
```

I don't usually watch the logs so close. Perhaps we should just throw an exception so the system won't start under this circumstance? Or can it ever be a valid case?

---

A fatal error sounds good to me :)
Perhaps the name of config could be included in the error message as well for easier debugging?